### PR TITLE
fix(agent): support FlowAgent sub-agents in routing agent

### LINKED ISSUE #4553: [Feature] Support FlowAgent (SequentialAgent/ParallelAgent) as subAgents in LlmRoutingAgent
### Feature Request

希望 LlmRoutingAgent 的 subAgents 支持 FlowAgent 嵌套，实现 Router + Workflow + Agent 的多层编排能力。

### 错误
尝试将 SequentialAgent 作为 LlmRoutingAgent 的 subAgent 时抛出异常。
#### 代码
```
@Bean
public LlmRoutingAgent nimbusRouterAgent(ChatModel chatModel,
                                          SequentialAgent weatherSequentialAgent,
                                          ReactAgent chitchatAgent) {
    return LlmRoutingAgent.builder()
            .name("router")
            .model(chatModel)
            .description("路由")
            .subAgents(List.of(weatherSequentialAgent, chitchatAgent))
            .build();
}
```
#### 异常
```
java.lang.IllegalArgumentException: Routing sub-agents must be BaseAgent for merge support
    at com.alibaba.cloud.ai.graph.agent.flow.strategy.RoutingGraphBuildingStrategy.buildCoreGraph(RoutingGraphBuildingStrategy.java:84)
```

### 原因
RoutingGraphBuildingStrategy 第 80-84 行硬性检查 instanceof BaseAgent，SequentialAgent 和 ParallelAgent 继承自 FlowAgent，不属于 BaseAgent 体系，因此被拒绝。
```
for (Agent subAgent : config.getSubAgents()) {
    if (!(subAgent instanceof BaseAgent)) {
        throw new IllegalArgumentException("Routing sub-agents must be BaseAgent for merge support");
    }
    baseAgentList.add((BaseAgent) subAgent);
}
```
### 环境信息
- spring-ai-alibaba-agent-framework: 1.1.2.2
- spring-ai: 1.1.4

### Describe the solution you'd like

希望 LlmRoutingAgent 的 subAgents 支持 FlowAgent（SequentialAgent、ParallelAgent 等），
使 Router 能挂载任意类型的 Agent，实现路由 + 工作流的嵌套编排。

期望用法：
```
LlmRoutingAgent router = LlmRoutingAgent.builder()
    .name("router")
    .model(chatModel)
    .subAgents(List.of(sequentialAgent1, sequentialAgent2))
    .build();
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jijiutong**
## Summary
- allow routing agents to accept FlowAgent sub-agents instead of only BaseAgent instances
- resolve routed FlowAgent outputs from nested final outputs, parallel merge outputs, or subgraph wrapper outputs
- add regression coverage for routing to a SequentialAgent workflow and merge behavior

Fixes #4553

## Tests
- ./mvnw -pl :spring-ai-alibaba-agent-framework -am -Dtest=FlowAgentArchitectureTest,RoutingMergeNodeTest,LlmRoutingFlowAgentIntegrationTest test

### CODE REVIEW

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java [hunk 1]
```diff
@@ -107,6 +121,47 @@ public Map<String, Object> apply(OverAllState state) throws Exception {
...
+			if (nestedAgents != null && !nestedAgents.isEmpty()) {
+				outputKeys.addAll(resolveOutputKeys(nestedAgents.get(nestedAgents.size() - 1),
+						allowDefaultMessagesOutput));
+			}
+		}
+		else if (agent instanceof FlowAgent flowAgent) {
+			List<Agent> nestedAgents = flowAgent.subAgents();
+			if (nestedAgents != null) {
+				for (Agent nestedAgent : nestedAgents) {
+					outputKeys.addAll(resolveOutputKeys(nestedAgent, allowDefaultMessagesOutput));
```
[the remark this answers is not in the document]
Comment (jijiutong) [reply spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java]: 👍

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java [hunk 2]
```diff
@@ -107,6 +154,146 @@ public Map<String, Object> apply(OverAllState state) throws Exception {
...
+						outputKeys.addAll(resolveOutputKeys(nestedAgent, false, allowRoutingMergedOutput, false));
+					}
+				}
+			}
+		}
+		else if (agent instanceof LlmRoutingAgent) {
+			if (allowRoutingMergedOutput) {
+				outputKeys.add(DEFAULT_MERGED_OUTPUT_KEY);
+			}
+		}
```
[the remark this answers is not in the document]
Comment (jijiutong) [reply spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java]: 👍

**Comment by jijiutong** [comment]
I checked a few other failing runs and it looks like they are failing on the same two tests:

- AgentToolNodeAsyncExecutionTest$StateUpdateOnTimeoutTests.stateUpdates_shouldBeCleared_onTimeoutWithCancellableTool
- CancellableAsyncToolCallbackTest$TimeoutScenarioTests.tool_shouldStopGracefully_whenCheckingCancellation

These tests don't seem to touch the routing code changed in this PR. They are about async tool cancellation, while this PR changes RoutingMergeNode / RoutingGraphBuildingStrategy.

I also tried running the related tests directly locally and they passed, but the full module test suite can still hit the same failure. So this looks more like an existing flaky test in agent-framework than a regression from this PR.

I don't have permission to rerun the failed job. Could someone rerun it, or let me know if you'd prefer a separate PR to make those cancellation tests more stable?

**Review by oss-sentinel-ai — COMMENTED** [review]
### Summary
Reviewed the changes in this PR (2704 lines changed). The changes appear to be well-structured.

### Observations
- Changes span 2704 lines
- Please ensure all edge cases are covered
- Verify test coverage for the modified code paths

---
*Automated review by github-manager-bot*

**Comment by oss-sentinel-ai** [comment]
⚠️ **Merge conflict detected**

This PR has conflicts with the `main` branch and cannot be merged. Please rebase or merge `main` into your branch and resolve the conflicts:

```bash
git fetch origin
git checkout fix-routing-flow-agent-subagents
git rebase origin/main
# resolve conflicts, then:
git push --force-with-lease
```

This is a one-time reminder. Feel free to @mention me for a re-review after conflicts are resolved.

---
*Automated notification by github-manager-bot*

File: spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/executor/NodeExecutor.java [hunk 3]
```diff
@@ -341,11 +342,15 @@ else if (element instanceof GraphResponse) {
 						}
 
 						if (result instanceof Map resultMap) {
-							if (!resultMap.containsKey(key) && resultMap.containsKey("messages")) {
-								List<Object> messages = (List<Object>) resultMap.get("messages");
-								Object lastMessage = messages.get(messages.size() - 1);
-								if (lastMessage instanceof AssistantMessage lastAssistantMessage) {
-									resultMap.put(key, lastAssistantMessage.getText());
+							if ((!resultMap.containsKey(key) || isSubGraphWrapperKey(key))
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/executor/NodeExecutor.java:345]: Streaming completion now overwrites an existing subgraph wrapper value with the last assistant message text. This changes behavior for every streaming subgraph (not just routing) and creates a streaming-vs-non-streaming inconsistency; scope it to routing wrappers or add broad compatibility coverage.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingOutputResolver.java [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+	 * @param preferWrapperOutput whether the wrapper key should be probed first
+	 * @return ordered output candidates
+	 */
+	private List<RoutingOutputCandidate> resolveCandidates(Agent agent, boolean allowDefaultMessagesOutput,
+			boolean allowRoutingMergedOutput, boolean preferWrapperOutput) {
+		List<RoutingOutputCandidate> candidates = new ArrayList<>();
+		Deque<RoutingOutputResolutionStep> steps = new ArrayDeque<>();
+		steps.push(new ExpandAgentOutputStep(agent, allowDefaultMessagesOutput, allowRoutingMergedOutput,
+				preferWrapperOutput));
+		while (!steps.isEmpty()) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingOutputResolver.java:240]: resolveCandidates walks agent sub-graphs without cycle detection; a custom FlowAgent that references itself (or forms a cycle) in subAgents will loop until stack overflow.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingOutputStrategy.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+	@Override
+	public void appendCandidates(ExpandAgentOutputStep step, Deque<RoutingOutputResolutionStep> steps,
+			List<RoutingOutputCandidate> candidates) {
+		BaseAgent baseAgent = (BaseAgent) step.agent();
+		String outputKey = baseAgent.getOutputKey();
+		if (outputKey != null) {
+			candidates.add(new RoutingOutputCandidate(outputKey, baseAgent, false));
+		}
+		else if (step.allowDefaultMessagesOutput()) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingOutputStrategy.java:91]: BaseAgentOutputStrategy now falls back to the shared messages key for a single routed BaseAgent without an outputKey. This changes simple (non-FlowAgent) routing behavior where such agents previously produced no result; confirm this is intended.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingOutputResolver.java [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+		if (isOpaqueCustomFlow(agent)) {
+			return true;
+		}
+		// A single workflow wrapper cannot attribute several nested routers because each
+		// router uses merged_result internally. Resolve their namespaced wrappers instead.
+		if (countRoutingMergedOutputProducers(agent) > 1) {
+			return false;
+		}
+
+		Set<String> outputKeys = resolvedNonWrapperOutputKeys(agent, routingMergedOutputProducerCount);
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingOutputResolver.java:366]: shouldPreferWrapperOutput recomputes resolvedNonWrapperOutputKeys for every sibling, repeating the same tree traversal; consider caching the key set per top-level agent.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java [hunk 7]
```diff
@@ -131,28 +188,237 @@ private String synthesize(String query, List<String> formattedResults) {
...
+		return result;
+	}
+
+	/**
+	 * Detects wrapper snapshot keys that belong to graph execution state rather than
+	 * agent-visible output.
+	 * @param key wrapper snapshot key
+	 * @return true when the key should not be treated as an answer
+	 */
+	private static boolean isInternalWrapperStateKey(String key) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java:377]: isInternalWrapperStateKey treats any key starting with '_' or ending with '_input' as internal wrapper state, so legitimate user output keys matching those patterns would be ignored by the single-visible-wrapper fallback.

File: spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/executor/NodeExecutor.java [hunk 8]
```diff
@@ -803,6 +813,12 @@ private static String graphFluxStateKey(GraphFlux<?> graphFlux) {
 		return StringUtils.hasText(graphFlux.getKey()) ? graphFlux.getKey() : "result";
 	}
 
+	private static boolean isSubGraphWrapperKey(String outputKey) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/executor/NodeExecutor.java:816]: isSubGraphWrapperKey duplicates the wrapper-key predicate already present in RoutingMergeNode and ResumableSubGraphAction; centralize it to avoid divergence if the naming convention changes.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java [hunk 9]
```diff
@@ -131,28 +188,237 @@ private String synthesize(String query, List<String> formattedResults) {
...
+		}
+		return false;
+	}
+
+	/**
+	 * Detects the parent-state wrapper key used for embedded subgraph results.
+	 * @param outputKey candidate output key
+	 * @return true if the key is a subgraph wrapper key
+	 */
+	private static boolean isSubGraphWrapperKey(String outputKey) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/flow/node/RoutingMergeNode.java:409]: isSubGraphWrapperKey duplicates the wrapper-key predicate used in NodeExecutor; centralize the check via ResumableSubGraphAction.

**Review by oss-sentinel-ai — CHANGES_REQUESTED** [review]
﻿### Summary
Re-review after 6 new commits since my last review (Aug 11). This PR extends `LlmRoutingAgent` so routing targets can be any `Agent` (Sequential/Parallel/Loop/custom FlowAgent/nested routers), backed by a new output-resolution layer (`RoutingOutputResolver` / `RoutingOutputCandidate` / `RoutingOutputStrategy`) and very extensive tests (+2,100 lines). The routing-side design is solid, and the recent hardening commits address real edge cases (namespaced selection markers, stale-wrapper cleanup, opaque custom flows). However, the `NodeExecutor` change in `graph-core` is a broad behavior change in shared runtime code that goes beyond the routing use case and currently lacks test coverage there — requesting changes before approval.

### Findings
- **[Critical]** `NodeExecutor.java:345` — the streaming completion path now overwrites an existing `subgraph_*_compiled_graph` wrapper value with the last assistant message text; this affects every streaming subgraph, not just routing, and creates a streaming-vs-non-streaming inconsistency.
- **[Warning]** `RoutingOutputResolver.java:240` — candidate resolution walks agent sub-graphs without a cycle guard.
- **[Warning]** `RoutingOutputStrategy.java:91` — simple (non-FlowAgent) routing behavior changes for routed `BaseAgent`s without an explicit `outputKey`.
- **[Info]** `RoutingOutputResolver.java:366`, `RoutingMergeNode.java:377`, `NodeExecutor.java:816`, `RoutingMergeNode.java:409` — see inline comments.

### Suggestions
1. **Scope or justify the NodeExecutor streaming overwrite.** Consider applying the wrapper overwrite only when the execution is a routing merge (or when the wrapper was cleared this turn), or document why all streaming subgraphs should have their wrapper value replaced by plain text. Either way, add a test covering a non-routing streaming subgraph so the shared-runtime behavior change is explicit.
2. **Guard against cycles in `resolveCandidates`.** A custom FlowAgent whose `subAgents` reference itself (directly or indirectly) would loop until stack overflow; a visited-set per resolution, or an explicit "agent graphs must be acyclic" contract check, would be cheap insurance.
3. **Centralize the `subgraph_*_compiled_graph` predicate.** `isSubGraphWrapperKey` is now duplicated in `NodeExecutor` and `RoutingMergeNode` (and conceptually in `ResumableSubGraphAction`); exposing one shared constant/helper avoids silent drift.

### Tests
Unit and integration coverage of the new resolver, stale-wrapper handling, marker isolation and shared-key attribution is strong. `NodeExecutorTest` currently only covers the wrapper-snapshot exclusion — the streaming overwrite path deserves coverage too.

---
*Automated review by github-manager-bot*
